src/dataflow/components/dataflow-program.tsx

```diff
@@ -372,6 +383,27 @@ export class DataflowProgram extends BaseComponent<IProps, IState> {
     })();
   }
 
+  private keepNodesInView = () => {
+    const margin = 5;
+    const { k } = this.programEditor.view.area.transform;
+    let rightMostEdge = 0;
+    let bottomMostEdge = 0;
+    this.programEditor.nodes.forEach((n: Node) => {
+      const nodeView = this.programEditor.view.nodes.get(n);
+      if (nodeView) {
+        rightMostEdge = Math.max (rightMostEdge, (nodeView.node.position[0] + nodeView.el.clientWidth) * k);
+        bottomMostEdge = Math.max (bottomMostEdge, (nodeView.node.position[1] + nodeView.el.clientHeight) * k);
+      }
+    });
```

Comment:
I'd be inclined to pull this piece out as a separately callable function -- something like `getProgramBounds()` or `getBoundingRectOfNodes()`.

Reply:
Agreed.  Added `getBoundingRectOfNodes` function.